File: java/client-apis/src/main/java/org/apache/rocketmq/client/apis/ClientConfigurationBuilder.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.rocketmq.client.apis;

import static com.google.common.base.Preconditions.checkArgument;
import static com.google.common.base.Preconditions.checkNotNull;

import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.regex.Pattern;
import org.apache.rocketmq.client.apis.consumer.SimpleConsumer;
import org.apache.rocketmq.client.apis.consumer.SimpleConsumerBuilder;

/**
 * Builder to set {@link ClientConfiguration}.
 */
public class ClientConfigurationBuilder {
    private static final int MAX_CLIENT_PROPERTIES_ENTRIES = 32;
    private static final int MAX_CLIENT_PROPERTY_KEY_LENGTH = 64;
    private static final int MAX_CLIENT_PROPERTY_VALUE_LENGTH = 256;
    private static final int MAX_CLIENT_PROPERTIES_TOTAL_SIZE_BYTES = 4 * 1024;
    private static final String RESERVED_CLIENT_PROPERTY_PREFIX = "rocketmq.";
    private static final Pattern CLIENT_PROPERTY_KEY_PATTERN = Pattern.compile("[a-zA-Z][a-zA-Z0-9_.-]*");

    private String endpoints;
    private SessionCredentialsProvider sessionCredentialsProvider = null;
    private Duration requestTimeout = Duration.ofSeconds(3);
    private boolean sslEnabled = true;
    private String namespace = "";
    private int maxStartupAttempts = 3;
    private final Map<String, String> clientProperties = new LinkedHashMap<>();

    /**
     * Configure the access point with which the SDK should communicate.
     *
     * @param endpoints address of service.
     * @return the client configuration builder instance.
     */
    public ClientConfigurationBuilder setEndpoints(String endpoints) {
        checkNotNull(endpoints, "endpoints should not be null");
        this.endpoints = endpoints;
        return this;
    }

    /**
     * Config the session credential provider.
     *
     * @param sessionCredentialsProvider session credential provider.
     * @return the client configuration builder instance.
     */
    public ClientConfigurationBuilder setCredentialProvider(SessionCredentialsProvider sessionCredentialsProvider) {
        this.sessionCredentialsProvider = checkNotNull(sessionCredentialsProvider, "credentialsProvider should not " +
            "be null");
        return this;
    }

    /**
     * Configure request timeout for ordinary RPC.
     *
     * <p>request timeout is 3s by default. Especially, the RPC request timeout for long-polling of
     * {@link SimpleConsumer} is increased by request timeout here based on the
     * {@linkplain SimpleConsumerBuilder#setAwaitDuration(Duration) await duration}.
     *
     * @param requestTimeout RPC request timeout.
     * @return the client configuration builder instance.
     */
    public ClientConfigurationBuilder setRequestTimeout(Duration requestTimeout) {
        this.requestTimeout = checkNotNull(requestTimeout, "requestTimeout should not be null");
        return this;
    }

    /**
     * Enable or disable the use of Secure Sockets Layer (SSL) for network transport.
     *
     * @param sslEnabled A boolean value indicating whether SSL should be enabled or not.
     * @return The {@link ClientConfigurationBuilder} instance, to allow for method chaining.
     */
    public ClientConfigurationBuilder enableSsl(boolean sslEnabled) {
        this.sslEnabled = sslEnabled;
        return this;
    }

    /**
     * Configure namespace for client
     * @param namespace namespace
     * @return The {@link ClientConfigurationBuilder} instance, to allow for method chaining.
     */
    public ClientConfigurationBuilder setNamespace(String namespace) {
        this.namespace = checkNotNull(namespace, "namespace should not be null");
        return this;
    }

    /**
     * Configure maxStartupAttempts for client
     *
     * @param maxStartupAttempts max attempt times when client startup
     * @return The {@link ClientConfigurationBuilder} instance, to allow for method chaining.
     */
    public ClientConfigurationBuilder setMaxStartupAttempts(int maxStartupAttempts) {
        checkArgument(maxStartupAttempts > 0, "maxStartupAttempts should more than 0");
        this.maxStartupAttempts = maxStartupAttempts;
        return this;
    }

    /**
     * Add a client instance property reported to server-side client runtime.
     *
     * @param key property key.
     * @param value property value.
     * @return the client configuration builder instance.
     */
    public ClientConfigurationBuilder addClientProperty(String key, String value) {
        validateClientPropertyEntry(key, value);
        Map<String, String> candidate = new LinkedHashMap<>(clientProperties);
        candidate.put(key, value);
        validateClientPropertiesLimits(candidate);
        this.clientProperties.clear();
        this.clientProperties.putAll(candidate);
        return this;
    }

    /**
     * Set client instance properties reported to server-side client runtime, replacing existing properties.
     *
     * @param properties client properties.
     * @return the client configuration builder instance.
     */
    public ClientConfigurationBuilder setClientProperties(Map<String, String> properties) {
        checkNotNull(properties, "clientProperties should not be null");
        Map<String, String> candidate = new LinkedHashMap<>(properties);
        validateClientProperties(candidate);
        this.clientProperties.clear();
        this.clientProperties.putAll(candidate);
        return this;
    }

    /**
     * Remove a client instance property.
     *
     * @param key property key.
     * @return the client configuration builder instance.
     */
    public ClientConfigurationBuilder removeClientProperty(String key) {
        checkNotNull(key, "client property key should not be null");
        this.clientProperties.remove(key);
        return this;
    }

    /**
     * Clear all client instance properties.
     *
     * @return the client configuration builder instance.
     */
    public ClientConfigurationBuilder clearClientProperties() {
        this.clientProperties.clear();
        return this;
    }

    /**
     * Finalize the build of {@link ClientConfiguration}.
     *
     * @return the client configuration builder instance.
     */
    public ClientConfiguration build() {
        checkNotNull(endpoints, "endpoints should not be null");
        checkNotNull(requestTimeout, "requestTimeout should not be null");
        // Keep build() defensive for maps supplied through setClientProperties or future builder paths.
        validateClientProperties(clientProperties);
        return new ClientConfiguration(endpoints, sessionCredentialsProvider, requestTimeout, sslEnabled, namespace,
            maxStartupAttempts, clientProperties);
    }

    private static void validateClientProperties(Map<String, String> properties) {
        validateClientProperties(properties, true);
    }

    private static void validateClientPropertiesLimits(Map<String, String> properties) {
        validateClientProperties(properties, false);
    }

    /**
     * Performs a client-side pre-check for client properties before the settings are reported to the server.
     *
     * <p>This validation mirrors the public proto constraints as closely as possible without depending on generated
     * proto classes in the client-apis module. The server still has final authority and may reject the properties with
     * its own validation result.
     */
    private static void validateClientProperties(Map<String, String> properties, boolean validateEntries) {
        checkArgument(properties.size() <= MAX_CLIENT_PROPERTIES_ENTRIES,
            "clientProperties should not contain more than %s entries", MAX_CLIENT_PROPERTIES_ENTRIES);
        int totalSize = 0;
        for (Map.Entry<String, String> entry : properties.entrySet()) {
            String key = entry.getKey();
            String value = entry.getValue();
            if (validateEntries) {
                validateClientPropertyEntry(key, value);
            }
            totalSize += computeClientPropertySize(key, value);
        }
        checkArgument(totalSize <= MAX_CLIENT_PROPERTIES_TOTAL_SIZE_BYTES,
            "clientProperties total size should not exceed %s bytes", MAX_CLIENT_PROPERTIES_TOTAL_SIZE_BYTES);
    }

    private static void validateClientPropertyEntry(String key, String value) {
        checkNotNull(key, "client property key should not be null");
        checkNotNull(value, "client property value should not be null");
        checkArgument(!key.isEmpty(), "client property key should not be empty");
        checkArgument(key.length() <= MAX_CLIENT_PROPERTY_KEY_LENGTH,
            "client property key length should not exceed %s characters", MAX_CLIENT_PROPERTY_KEY_LENGTH);
        checkArgument(CLIENT_PROPERTY_KEY_PATTERN.matcher(key).matches(),
            "client property key should start with a letter and only contain letters, digits, dot, underscore or "
                + "hyphen");
        checkArgument(!key.startsWith(RESERVED_CLIENT_PROPERTY_PREFIX),
            "client property key should not use reserved prefix %s", RESERVED_CLIENT_PROPERTY_PREFIX);
        checkArgument(value.length() <= MAX_CLIENT_PROPERTY_VALUE_LENGTH,
            "client property value length should not exceed %s characters", MAX_CLIENT_PROPERTY_VALUE_LENGTH);
    }

    /**
     * Computes the serialized size of one {@code client_properties} map entry in protobuf wire format.
     *
     * <p>A {@code map<string, string>} field is encoded as repeated entry messages. Each entry is written as Settings
     * field 9, containing key field 1 and value field 2. Client property keys are constrained to ASCII, so
     * {@link String#length()} equals the UTF-8 byte size for keys; values may contain non-ASCII characters and must use
     * their UTF-8 byte size.
     */
    private static int computeClientPropertySize(String key, String value) {
        int keySize = key.length();
        int valueSize = value.getBytes(StandardCharsets.UTF_8).length;
        int entrySize = computeTagSize(1)
            + computeUInt32SizeNoTag(keySize)
            + keySize
            + computeTagSize(2)
            + computeUInt32SizeNoTag(valueSize)
            + valueSize;
        return computeTagSize(9)
            + computeUInt32SizeNoTag(entrySize)
            + entrySize;
    }

    /**
     * Computes the protobuf wire-format size of a field tag for the given field number.
     */
    private static int computeTagSize(int fieldNumber) {
        return computeUInt32SizeNoTag(fieldNumber << 3);
    }

    /**
     * Computes the protobuf unsigned int32 varint size.
     */
    private static int computeUInt32SizeNoTag(int value) {
        if ((value & (~0 << 7)) == 0) {
            return 1;
        }
        if ((value & (~0 << 14)) == 0) {
            return 2;
        }
        if ((value & (~0 << 21)) == 0) {
            return 3;
        }
        if ((value & (~0 << 28)) == 0) {
            return 4;
        }
        return 5;
    }
}

File: php/ClientConfigurationBuilder.php
<?php
/**
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

namespace Apache\Rocketmq;

/**
 * ClientConfigurationBuilder - Fluent builder for ClientConfiguration.
 * Referencing Java ClientConfigurationBuilder.java
 */
class ClientConfigurationBuilder
{
    private ?string $endpoints = null;
    private ?SessionCredentials $sessionCredentialsProvider = null;
    private int $requestTimeoutMs = 3000; // default 3s
    private bool $sslEnabled = true;
    private string $namespace = '';
    private int $maxStartupAttempts = 3;
    private ?TlsCredentials $tlsCredentials = null;

    /**
     * Set the service endpoint address (required).
     *
     * @param string $endpoints e.g. "127.0.0.1:8080"
     * @return $this
     */
    public function setEndpoints(string $endpoints): self
    {
        $this->endpoints = $endpoints;
        return $this;
    }

    /**
     * Set authentication credentials provider.
     *
     * @param SessionCredentials|null $credentials
     * @return $this
     */
    public function setCredentialProvider(?SessionCredentials $credentials): self
    {
        $this->sessionCredentialsProvider = $credentials;
        return $this;
    }

    /**
     * Set RPC request timeout.
     *
     * @param int $timeoutMs Timeout in milliseconds (default 3000)
     * @return $this
     * @throws \InvalidArgumentException if timeout is zero or negative
     */
    public function setRequestTimeout(int $timeoutMs): self
    {
        if ($timeoutMs <= 0) {
            throw new \InvalidArgumentException("Request timeout must be > 0");
        }
        $this->requestTimeoutMs = $timeoutMs;
        return $this;
    }

    /**
     * Enable or disable SSL.
     *
     * @param bool $enabled Default true
     * @return $this
     */
    public function enableSsl(bool $enabled = true): self
    {
        $this->sslEnabled = $enabled;
        return $this;
    }

    /**
     * Set namespace.
     *
     * @param string $namespace Default empty
     * @return $this
     */
    public function setNamespace(string $namespace): self
    {
        $this->namespace = $namespace;
        return $this;
    }

    /**
     * Set max startup retry attempts.
     *
     * @param int $attempts Must be > 0
     * @return $this
     * @throws \InvalidArgumentException if attempts is zero or negative
     */
    public function setMaxStartupAttempts(int $attempts): self
    {
        if ($attempts <= 0) {
            throw new \InvalidArgumentException("Max startup attempts must be > 0");
        }
        $this->maxStartupAttempts = $attempts;
        return $this;
    }

    /**
     * Set TLS credentials for the gRPC connection.
     *
     * @param TlsCredentials $tlsCredentials
     * @return $this
     */
    public function setTlsCredentials(TlsCredentials $tlsCredentials): self
    {
        $this->tlsCredentials = $tlsCredentials;
        $this->sslEnabled = true;
        return $this;
    }

    /**
     * Enable TLS with default system CA bundle.
     *
     * @return $this
     */
    public function enableTls(): self
    {
        $this->tlsCredentials = TlsCredentials::createDefault();
        $this->sslEnabled = true;
        return $this;
    }

    /**
     * Enable mutual TLS (mTLS) with client certificate and key.
     *
     * @param string $clientCertPath Path to client certificate file
     * @param string $clientKeyPath Path to client private key file
     * @param string|null $caCertPath Optional CA certificate path
     * @return $this
     */
    public function enableMutualTls(
        string $clientCertPath,
        string $clientKeyPath,
        ?string $caCertPath = null
    ): self {
        $this->tlsCredentials = TlsCredentials::createMtls(
            $clientCertPath,
            $clientKeyPath,
            $caCertPath
        );
        $this->sslEnabled = true;
        return $this;
    }

    /**
     * Enable TLS but skip peer verification (development only).
     *
     * ⚠️ SECURITY WARNING: This method is for internal testing use ONLY.
     * It should NOT be used in production code as it allows man-in-the-middle attacks.
     *
     * For development/testing, use environment variables or configuration files instead:
     * - Set ROCKETMQ_TLS_SKIP_VERIFY=true in your .env file
     * - Use TlsCredentials::createInsecureDev() directly with explicit awareness of risks
     *
     * @return $this
     * @internal Not intended for public API usage
     * @deprecated Will be removed in future versions. Use environment-based configuration instead.
     */
    public function disableTlsVerification(): self
    {
        trigger_error(
            "SECURITY WARNING: ClientConfigurationBuilder::disableTlsVerification() is deprecated and will be removed. " .
            "This method bypasses TLS certificate verification and allows man-in-the-middle attacks. " .
            "For development/testing, use TlsCredentials::createInsecureDev() explicitly with full awareness of security risks. " .
            "NEVER use this in production!",
            E_USER_DEPRECATED
        );
        
        Logger::getInstance('ClientConfiguration')->error(
            "CRITICAL SECURITY WARNING: TLS certificate verification is being disabled! " .
            "This makes the connection vulnerable to man-in-the-middle attacks. " .
            "This should ONLY be used in isolated development/testing environments. " .
            "DO NOT use this in production under any circumstances!"
        );
        
        $this->tlsCredentials = TlsCredentials::createInsecureDev();
        $this->sslEnabled = true;
        return $this;
    }

    /**
     * Build the immutable ClientConfiguration object.
     *
     * @return ClientConfiguration
     * @throws \InvalidArgumentException if endpoints not set
     */
    public function build(): ClientConfiguration
    {
        if ($this->endpoints === null || $this->endpoints === '') {
            throw new \InvalidArgumentException("Endpoints must be set");
        }
        if ($this->requestTimeoutMs <= 0) {
            throw new \InvalidArgumentException("Request timeout must be > 0");
        }

        return ClientConfiguration::create(
            $this->endpoints,
            $this->sessionCredentialsProvider,
            $this->requestTimeoutMs,
            $this->sslEnabled,
            $this->namespace,
            $this->maxStartupAttempts,
            $this->tlsCredentials
        );
    }
}
